Deleting an environment fails on Windows

### Terraform Version
Terraform v0.9.8

### Debug Output
Pasted here as it is very short:

```
2017/06/20 15:35:27 [WARN] Invalid log level: "1". Defaulting to level: TRACE. Valid levels are: [TRACE DEBUG INFO WARN ERROR]
2017/06/20 15:35:28 [INFO] Terraform version: 0.9.8  8d560482c34e865458fd884cb0790b4f73f09ad1
2017/06/20 15:35:28 [INFO] Go runtime version: go1.8
2017/06/20 15:35:28 [INFO] CLI args: []string{"C:\\Users\\vjimenez\\scoop\\apps\\terraform\\current\\terraform.exe", "env", "delete", "xxx"}
2017/06/20 15:35:28 [DEBUG] Attempting to open CLI config file: C:\Users\vjimenez\AppData\Roaming\terraform.rc
2017/06/20 15:35:28 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2017/06/20 15:35:28 [INFO] CLI command args: []string{"env", "delete", "xxx"}
2017/06/20 15:35:28 [DEBUG] command: loading backend config file: E:\test\terraform-env-test\test
2017/06/20 15:35:28 [INFO] command: backend config not found, returning nil: E:\test\terraform-env-test\test
2017/06/20 15:35:28 [DEBUG] command: no data state file found for backend config
2017/06/20 15:35:28 [DEBUG] New state was assigned lineage "95d13254-4b4b-4114-aec7-7ce92a0acb52"
2017/06/20 15:35:28 [INFO] command: backend initialized: <nil>
2017/06/20 15:35:28 [INFO] command: backend <nil> is not enhanced, wrapping in local
2017/06/20 15:35:28 [DEBUG] plugin: waiting for all plugin processes to complete...
remove terraform.tfstate.d\xxx\terraform.tfstate: The process cannot access the file because it is being used by another process.
```

### Expected Behavior
The environment should be deleted.

### Actual Behavior
The environment is not deleted and a wrong error is given. File `terraform.tfstate.d\xxx\terraform.tfstate` cannot be used by another process as it simply does not exist (`terraform env new` does not create any state file).

### Steps to Reproduce
1. `terraform env new xxx`
2. `terraform env select default`
3. `terraform env delete xxx`

### Important Factoids
The problem only occurs when running on Windows (the same steps on Linux work as expected). I am using Windows 10 (version 1703).

After more investigation, it seems the problem is related to a particular behavior in Windows.

While the following code (creating, deleting and closing a file) works in Linux, in Windows I get the same error that I get when trying to delete a Terraform workspace.

```go
package main

import (
  "fmt"
  "os"
)

func main() {
  file, err := os.Create("test.txt")
  if err != nil {
    panic(fmt.Sprintf("Error opening file: %s", err))
  }

  err = os.Remove("test.txt")
  if err != nil {
    panic(fmt.Sprintf("Error deleting file: %s", err))
  }

  err = file.Close()
  if err != nil {
    panic(fmt.Sprintf("Error closing file: %s", err))
  }
}
```

Error message:
```
panic: Error deleting file: remove test.txt: The process cannot access the file because it is being used by another process.
```

--- a/command/workspace_delete.go
+++ b/command/workspace_delete.go
@@ -105,6 +105,7 @@ func (c *WorkspaceDeleteCommand) Run(args []string) int {
 		return 1
 	}
 
+	// Honor the lock request, for consistency and one final safety check.
 	if c.stateLock {
 		lockCtx, cancel := context.WithTimeout(context.Background(), c.stateLockTimeout)
 		defer cancel()
@@ -117,7 +118,17 @@ func (c *WorkspaceDeleteCommand) Run(args []string) int {
 			c.Ui.Error(fmt.Sprintf("Error locking state: %s", err))
 			return 1
 		}
-		defer clistate.Unlock(sMgr, lockID, c.Ui, c.Colorize())
+
+		// We need to release the lock just before deleting the state, in case
+		// the backend can't remove the resource while holding the lock. This
+		// is currently true for Windows local files.
+		//
+		// TODO: While there is little safety in locking while deleting the
+		// state, it might be nice to be able to coordinate processes around
+		// state deletion, i.e. in a CI environment. Adding Delete() as a
+		// required method of States would allow the removal of the resource to
+		// be delegated from the Backend to the State itself.
+		clistate.Unlock(sMgr, lockID, c.Ui, c.Colorize())
 	}
 
 	err = b.DeleteState(delEnv)

